fix #8970: remove space between accidentals and arpeggios

Nick-Mazuk opened the pull request:
Resolves: https://github.com/musescore/MuseScore/issues/8970

Accidentals have a quarter space padding around them. This creates an unnecessarily large gap between accidentals and arpeggios. This PR decreases that gap by a quarter space, making the gap the same size as if the accidentals weren't there.

Before:

<img width="709" alt="Screen Shot 2021-09-09 at 4 00 08 PM" src="https://user-images.githubusercontent.com/5659171/132773440-a8d06a86-9749-450c-b35d-1b76f9f54312.png">

After:

<img width="707" alt="Screen Shot 2021-09-09 at 3 55 52 PM" src="https://user-images.githubusercontent.com/5659171/132773100-fa727505-6824-4877-bdd0-ebaef0270498.png">

Currently, the decrease of a quarter space is hardcoded in. Alternatively, we can use the same value that defines the spaces between accidentals. I have not yet found where this is defined, though.

Nick-Mazuk commented:
I could go either way. I tried both and went with 0.25 spaces because then the gap size is identical with and without accidentals.

Jojo-Schmitz commented:
Fair enough. To me in the images from original issue's score the distance between barline and arpeggio is still too small with 0.25. But that's nothing this PR here touches on, not directly at least.

Nick-Mazuk commented:
Just added a commit that does a few things:

1. Shrinks gap to just 0.25 spaces
2. Removes the hard coded magic number for the accidental padding
3. Added a style setting for the gap size. It's currently not exposed to the user, but it can be in the future if we wish

Nick-Mazuk commented:
This is a complete rewrite of the feature, but I believe it's finalized.

1. Increases gap back to 0.5 spaces
2. Decreases the gap for ♯ by the [cutout](https://w3c.github.io/smufl/latest/specification/bbox-cut-outs.html) to make it visually consistent with ♭
3. The gap follows the contour of the arpeggio and accidentals (i.e., if an accidental can fit closer to the arpeggio than it's bounding box, it will be closer)

For instance, take a look at this image. The sharp now fits inside of the bracket because there's space!

![image](https://user-images.githubusercontent.com/5659171/133491882-fe4048b4-52a2-44b1-945e-fb4e8afdf9c1.png)

All test cases:

![accidental-test](https://user-images.githubusercontent.com/5659171/133492759-8efb541c-8208-44b9-97e6-420c4eadb679.png)

Here's the test file I used to validate the changes.

[accidental-test.mscz.zip](https://github.com/musescore/MuseScore/files/7172554/accidental-test.mscz.zip)

Nick-Mazuk commented:
> ; missing

Fixed

Nick-Mazuk commented:
Assuming this looks good with everyone, I'm done working on it and it can be merged.

Jojo-Schmitz commented:
You may want to squash commits. Those last 3 commits should really be just one

Jojo-Schmitz commented on src/engraving/libmscore/chord.cpp:
```diff
@@ -1914,6 +1914,12 @@ void Chord::layoutPitched()
     //  process notes
     //-----------------------------------------
 
+    // Keeps track if there are any accidentals in this cord.
```
typo, "chord"

Nick-Mazuk commented:
Fixed typo and squashed the commits

Nick-Mazuk commented:
@Jojo-Schmitz, resolved your comments about the unreachable `breaks` and qreals.

Nick-Mazuk commented:
As requested by @oktophonie, the gap at measure 15 has been reduced to this:

<img width="353" alt="Screen Shot 2021-09-19 at 8 45 46 PM" src="https://user-images.githubusercontent.com/5659171/133955221-7b9ff013-b4b5-48c4-9c3a-441d626ac2be.png">

Nick-Mazuk commented:
These last few updates take care of every last edge case I can think of, including checking the output in multiple fonts. There are also a few minor tweaks as requested by Simon.

There's also some code refactoring to make sure it's easier to read.

Here's what the original test cases look like now:

![accidental-test](https://user-images.githubusercontent.com/5659171/134219592-977a6951-ee2b-4663-8da9-696b3aafa953.png)

And with Petaluma, a font with completely different metrics:

![accidental-test](https://user-images.githubusercontent.com/5659171/134219896-ab44ff04-ccda-4edf-82bc-1b9e7bba3aab.png)

vpereverzev commented on src/engraving/libmscore/arpeggio.cpp:
```diff
@@ -23,14 +23,17 @@
 #include "arpeggio.h"
 
 #include <cmath>
+#include <QVector>
```
At the moment we're trying to reduce dependencies from Qt in engraving module. Make sure that you're using STL containers here and in the further PRs

Nick-Mazuk replied:
Ok, will do. If I notice that there's an unneeded, existing Qt dependency, I assume I should convert it to a STL container?

Jojo-Schmitz replied:
In my backport of this PR to 3.x that include was not needed (Neither was the cmath) and 3.x's all.h doesn't either.

Jojo-Schmitz replied:
You're actually using a `QVector` in chord.cpp too bit not the include

DmitryArefiev commented:
Assigning to @vshalkevich

vshalkevich commented:
Checked on Linux Ubunty, Windows 10, MacOS
Works fine
